examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+#[init(contract = "CIS2-wCCD-State", enable_logger)]
+fn contract_state_init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+    logger: &mut impl HasLogger,
+) -> InitResult<State<S>> {
+    // Construct the initial contract state.
+    let state = State::new(state_builder);
+    // Get the instantiater of this contract instance.
+    let invoker = Address::Account(ctx.init_origin());
+    // Log event for the newly minted token.
+    logger.log(&Cis2Event::Mint(MintEvent {
+        token_id: TOKEN_ID_WCCD,
+        amount:   ContractTokenAmount::from(0u64),
+        owner:    invoker,
+    }))?;
+
+    // Log event for where to find metadata for the token
+    logger.log(&Cis2Event::TokenMetadata::<_, ContractTokenAmount>(TokenMetadataEvent {
+        token_id:     TOKEN_ID_WCCD,
+        metadata_url: MetadataUrl {
+            url:  String::from(TOKEN_METADATA_URL),
+            hash: None,
+        },
+    }))?;
+
+    Ok(state)
+}
+
+/// Initialzes the state of the w_ccd contract with the proxy and the
+/// implementation addresses.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "initialize",
+    parameter = "InitializeStateParams",
+    mutable
+)]
+fn contract_state_initialize<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<()> {
+    ensure!(!host.state().initialized, concordium_cis2::Cis2Error::Custom(AlreadyInitialized));
+    // Set proxy and implementation addresses.
+    let params: InitializeStateParams = ctx.parameter_cursor().get()?;
+    host.state_mut().proxy_address = params.proxy_address;
+    host.state_mut().implementation_address = params.implementation_address;
+    // Set the w_ccd_state contract as initialized.
+    host.state_mut().initialized = true;
+    Ok(())
+}
+
+// Getter and setter functions of state contract.
+
+fn only_implementation(
+    implementation: Option<ContractAddress>,
+    sender: Address,
+) -> ContractResult<()> {
+    match implementation {
+        Some(implementation_address) => {
+            ensure_eq!(
+                sender,
+                Address::Contract(implementation_address),
+                concordium_cis2::Cis2Error::Custom(OnlyImplementation)
+            );
+        }
+        None => bail!(concordium_cis2::Cis2Error::Custom(OnlyImplementation)),
+    };
+    Ok(())
+}
+
+/// Set unpause_time.
+#[receive(contract = "CIS2-wCCD-State", name = "set_unpause_time", mutable)]
+fn contract_state_set_unpause_time<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<()> {
+    // Only implementation can set state.
+    only_implementation(host.state().implementation_address, ctx.sender())?;
+
+    // Set unpause_time.
+    let unpause_time = ctx.parameter_cursor().get()?;
+    host.state_mut().unpause_time = unpause_time;
+    Ok(())
+}
```

Comment:
I guess here we don't want to have any opinions about a valid unpause time, e.g., that it be in the future?

Reply:
I'm not objecting to that, but it probably makes sense to add a bit of documentation for rationale.